### ISSUE TITLE
Using various source ports makes firewall rules difficult

### ISSUE THREAD

**Issue by NathanC** [mclarkk/lifxlan#114]
First off, awesome library! Thanks for writing it.

**tl;dr; Response to discovery from a lifx device is unicast to the source port, which is assigned by the OS. This makes it difficult to create a firewall rule allowing the response.** 

lan discovery doesn't work for me, due to my deny-all-incoming-by-default firewall rules. I allowed incoming UDP messages on 56700, but it still didn't work-- I had to allow a large range of incoming UDP ports, at which point discovery worked.

I captured my traffic on wireshark to investigate, and noticed that every discovery broadcast came from a different source port, and that my lightbulb ended up replying to that source port, instead of 56700. Investigating why that was, I stumbled across this from the [lifx-protocol-docs](https://github.com/LIFX/lifx-protocol-docs/blob/master/messages/device.md):
>As of LIFX Protocol V2, LIFX devices sending unicast message responses will reply to whichever port number the client has bound. See message header frame source identifier regarding the conditions under which unicast messages are sent.
>
> Prior to LIFX Protocol V2, i.e the Original LIFX A21 lightbulb running firmware version 1.x, LIFX devices send broadcast message responses to port 56700.
>
> To get the best compatibility across both newer and older LIFX devices, clients should bind to UDP port 56700.

Looking through the code, it seems like lifxlan allows the OS to assign a source port for the broadcast [in the initialize_socket method here.](https://github.com/mclarkk/lifxlan/blob/c429ac6d6b8e163ba848996e722645dff58b7ac6/lifxlan/lifxlan.py#L288)

I'd love to be able to specify which port is bound to for the broadcast (or a range), so that I can allow incoming UDP messages more selectively. Does that sound like a reasonable solution to you? If so, I could open a PR to expose those arguments to any consumers who wish to specify (with a default argument that leads to the existing behavior of an OS-selected port).

**Comment by mclarkk** [comment]
Originally the library only sent messages out of 56700. However, I believe that changed around the time that I added the Group functionality, where multiple sockets might need to be opened in parallel.

When I wrote that function, it was written [like this](https://github.com/mclarkk/lifxlan/commit/08cde0f9e91349be1c164bd73ea48ad50f9f0ffd#diff-243b8bde1615dce48ea609a4266afc07). If 56700 was in use it would simply increment and retry until it found a free port. It looks like a commit from @exking [changed that code](https://github.com/mclarkk/lifxlan/commit/3a13de5923db373b4ba23d3334dc2454119d49ea#diff-243b8bde1615dce48ea609a4266afc07) to allow the OS to automatically select the port.

So I have two questions:

1. @NathanC: If the socket code was changed back so that it would open port 56700 OR try the next ports, would that work better for your firewall rules? Can you allow a range of ports, say 56700 - 56720, or whatever, depending on the number of lights you have?

2. @exking Is there some bug you fixed by letting the OS choose the socket, or did you change it because you thought it was cleaner?

**Comment by exking** [comment]
In multi-threaded application (for example group command) - using a fixed port is not acceptable. Previous logic was just incrementing the port number, but 2 threads running simultaneous could be competing incrementing the port as well. 

I believe @NathanC's problem only happens during the initial discovery when broadcast packet is sent from a random port. Replies would come back from each individual bulb as a unicast, but his firewall is not letting them thru. Initial discovery should not really involve threading so using a fixed port for discovery only would be acceptable. (Another solution would be just to whitelist bulb IPs).

Later when unicast packets are sent to the bulbs - firewall should track the requests and let the replies back as "established" connection so using random OS assigned port for each request should not be a problem.

**Comment by NathanC** [comment]
Yeah, the issue is just during the unicast response to a broadcast during discovery, since firewalls don't seem to see that as an established connection. I think letting the OS select the port for the direct unicast -> unicast flows makes sense and firewalls generally [allow it as established](https://serverfault.com/questions/123208/iptables-p-udp-state-established#123212) as @exking pointed out.

I think the default behavior should stay as it is; if it's working for people overall now, there's no reason to change it. I'm personally in favor of just exposing an argument for discovery that allows a consumer to specify which port to try to bind to, if they so choose.

For my personal scripts, I'd just bind it to port 56700, which shouldn't conflict with anything on my system. For more robust code that still binds to a set of ports I've opened, I could recover from binding errors and keep looping until I find a port that works for me. 

e.g.,
```
def get_lights(self, specified_port = None):
     # propagate this down to the `initialize_socket` method.
     # if specified_port is None, then use a random port-- otherwise, 
     # attempt to bind to the specified port and raise an error if unable to
```
Or something like that. I'm pretty busy for the next week or so, but can open a PR when things settle down if the solution seems reasonable to you. There are other potential solutions, but this seems like a pretty non-invasive one; if people want to leverage it they can, but they don't have to.

**Comment by tacaswell** [comment]
I have a possibly silly protocol question: why keep creating new sockets?  Could the each {context, group, device} create a persistent socket that it re-uses?
